quickevent/app/plugins/Competitors/src/competitorswidget.cpp

```diff
@@ -217,43 +211,49 @@ void CompetitorsWidget::editCompetitor_helper(const QVariant &id, int mode, int
 {
 	qfLogFuncFrame() << "id:" << id << "mode:" << mode;
 	//qf::core::sql::Transaction transaction;
-	m_cbxEditCompetitorOnPunch->setEnabled(false);
-	auto *w = new CompetitorWidget();
-	w->setWindowTitle(tr("Edit Competitor"));
-	qfd::Dialog dlg(QDialogButtonBox::Save | QDialogButtonBox::Cancel, this);
-	dlg.setDefaultButton(QDialogButtonBox::Save);
-	bool save_and_next = false;
-	if(mode == qf::core::model::DataDocument::ModeInsert || mode == qf::core::model::DataDocument::ModeEdit) {
-		QPushButton *bt_save_and_next = dlg.buttonBox()->addButton(tr("Save and &next"), QDialogButtonBox::AcceptRole);
-		connect(dlg.buttonBox(), &qf::qmlwidgets::DialogButtonBox::clicked, [&save_and_next, bt_save_and_next](QAbstractButton *button) {
-			save_and_next = (button == bt_save_and_next);
-		});
-	}
-	dlg.setCentralWidget(w);
-	w->load(id, mode);
-	auto *doc = qobject_cast<Competitors::CompetitorDocument*>(w->dataController()->document());
-	QF_ASSERT(doc != nullptr, "Document is null!", return);
-	if(mode == qfm::DataDocument::ModeInsert) {
-		if(siid == 0) {
-			int class_id = m_cbxClasses->currentData().toInt();
-			doc->setValue("competitors.classId", class_id);
+	if(auto lock = std::unique_lock<std::mutex>{m_editCompetitorMutex, std::try_to_lock})
```

Comment:
the mutex is not necessary, what about
```c++
class EditGuard
{
public:
    EditGuard(bool &lock) : m_lock(lock) { m_lock = true; }
    ~EditGuard() { m_lock = false; }
private:
    bool &m_lock;
}
if(m_insideEditCompetitor) {
    qfDebug() << "Another competitor dialog is opened, ignoring..";
}
else {
    EditGuard g(m_insideEditCompetitor);
   ...
}
```

Reply:
why creating explicit class, when mutex is sufficient and designed for locking in the first place?

Reply:
Because purpose of Mutex is to synchronize threads. Using mutex in single-thread app can be confusing for contributors.

Reply:
ScopeGuard is c++ idiom for cases like that https://en.wikibooks.org/wiki/More_C%2B%2B_Idioms/Scope_Guard

Reply:
The Edit dialog can be trigerred by EditOnPunch, I thought there could be multiple threads in play, however I am not sure, how Qt application handles that events. If that is not a concern, its not needed, but I would still prefer mutex over additional class, just because its less code to maintain.

Also lock_guard or scoped_lock could be used for more clarity, but the need for manual unlocking lead me for unique_lock to avoid additional nesting.

Reply:
I think it's better to use Mutex than to write an explicit class. And we are ready for the future (which is multi-threaded) .

Reply:
> The Edit dialog can be trigerred by EditOnPunch, I thought there could be multiple threads in play, however I am not sure, how Qt application handles that events. If that is not a concern, its not needed, but I would still prefer mutex over additional class, just because its less code to maintain.

Whole QT UI must be run in single thread by QT design

Reply:
> I think it's better to use Mutex than to write an explicit class. And we are ready for the future (which is multi-threaded) .

not in QtWidgets

Reply:
now that i have cardreader on hand, I can say that no lock (class nor mutex) is needed, as the thread looking for punch is already blocked by the displayed dialog

Reply:
No it is not. QDialog just creates own QEventLoop object, what does mean that signals from timers or sockets or ... can be delivered meanwhile.

Reply:
To be honest I do not fully understand QT machinery yet. However, If you have edit competitor dialog opened, SI card readout dos not work (not even a beep from the box).

Reply:
try to insert this snippet before `dlg.exec()` in edit competitor
```c++
	QTimer tm;
	connect(&tm, &QTimer::timeout, []() {
		qfInfo() << "timeout";
	});
	tm.start(1000);
	bool ok = dlg.exec();
```
you will see "timeout" in log every second whilst dialog will be open

Reply:
ok, noted.
This still does not explain why, If you have edit competitor dialog opened, editcompetitor dialog is not reopened on repeated SI card punch, without any guard introduced. (I am curious about this, not that is an issue or point for necessary discussion)

If we assume a guard is needed anyway, then i still think mutex is a better solution, i would go further and say it is the only right solution as any other without a mutex is prone to race contidion and potentially locking the edit competitor dialog from opening. (That statement assumes all events from the app are not synchronized at some point and handled by one event loop)

Reply:
> This still does not explain why, If you have edit competitor dialog opened, editcompetitor dialog is not reopened on repeated SI card punch, without any guard introduced. (I am curious about this, not that is an issue or point for necessary discussion)

I guess that this is SQL concurrency issue and that this is rather bug. Qt itself allows concurrent SI reading while edit competitor dialog is open.

Reply:
> If we assume a guard is needed anyway, then i still think mutex is a better solution

hijacking mutex in single threaded application is always bad solution, it is also bad pattern for contributors and it makes code understanding harder. Pls, see our discussion above about Qt concurency, the mutex introducion introduces also many questions about rest of code.